src/core/Ekubo.sol

```diff
@@ -0,0 +1,453 @@
+    function locked(uint256 id) external returns (bytes memory);
+
+    /// @notice Called by Core on `msg.sender` to collect assets
+    /// @param id The id assigned to the action
+    /// @param token The token to pay on
+    function payCallback(uint256 id, address token) external;
+}
+
+library UnsafeEkuboCore {
+    function unsafeSwap(
+        IEkuboCore core,
+        PoolKey memory poolKey,
+        int128 amount,
+        bool isToken1,
+        SqrtRatio sqrtRatioLimit,
+        uint256 skipAhead
+    ) internal returns (int128 delta0, int128 delta1) {
+        assembly ("memory-safe") {
+            let ptr := mload(0x40)
+
+            mstore(ptr, 0x00000000) // selector for `swap_611415377((address,address,bytes32),int128,bool,uint96,uint256)`
+            // TODO: Check that native ETH is handled properly (0x0 vs 0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee)
+            mcopy(add(0x20, ptr), poolKey, 0x60)
+            mstore(add(0x80, ptr), signextend(0x0f, amount))
+            mstore(add(0xa0, ptr), isToken1)
+            mstore(add(0xc0, ptr), and(0xffffffffffffffffffffffff, sqrtRatioLimit))
+            mstore(add(0xe0, ptr), skipAhead)
+            
+            if iszero(call(gas(), core, 0x00, add(0x1c, ptr), 0xe4, 0x00, 0x40)) {
+                returndatacopy(ptr, 0x00, returndatasize())
+                revert(ptr, returndatasize())
+            }
+            delta0 := mload(0x00)
+            delta1 := mload(0x20)
+        }
+    }
+}
+
+abstract contract Ekubo is SettlerAbstract {
+    using UnsafeMath for uint256;
+    using UnsafeMath for int256;
+    using CreditDebt for int256;
+    using Ternary for bool;
+    using SafeTransferLib for IERC20;
+    using NotesLib for NotesLib.Note;
+    using NotesLib for NotesLib.Note[];
+    using StateLib for StateLib.State;
+    using UnsafeEkuboCore for IEkuboCore;
+
+    constructor() {
+        assert(BASIS == Encoder.BASIS);
+        assert(BASIS == Decoder.BASIS);
+        assert(ETH_ADDRESS == Decoder.ETH_ADDRESS);
+    }
+
+    // fill encoding musings:
+    //     bps (2 bytes)
+    //     packing key (1 byte)
+    //     sell/buy tokens (0, 20, or 40 bytes)
+    //     config (32 bytes) - (8 bytes fee, 4 bytes tickSpacing, 20 bytes extension)
+    //     skipAhead (32 bytes)
+
+    function sellToEkubo(
+        address recipient,
+        IERC20 sellToken,
+        uint256 bps,
+        bool feeOnTransfer,
+        uint256 hashMul,
+        uint256 hashMod,
+        bytes memory fills,
+        uint256 amountOutMin
+    ) internal returns (uint256 buyAmount) {
+        if (bps > BASIS) {
+            Panic.panic(Panic.ARITHMETIC_OVERFLOW);
+        }
+        bytes memory data = Encoder.encode(
+            uint32(IEkuboCore.lock.selector),
+            recipient,
+            sellToken,
+            bps,
+            feeOnTransfer,
+            hashMul,
+            hashMod,
+            fills,
+            amountOutMin
+        );
```

Comment:
You should verify that this produces the expected result even though `lock` takes an `uint256` argument instead of a `bytes` argument. I think that just means that `id` will always be 32. And an integration test will tell you for sure. But it's worth thinking over to make sure that there's nothing unexpected

Reply:
in this case lock takes no parameter, but got your point, double checking

Reply:
Ahh, I got confused about the callback ABI and the CORE ABI. So we can safely ignore the `id` argument to both of the callback functions?

Reply:
we either ignore them or use them for validation, we should receive the same id in all the callbacks, right now I am not doing any checks on that.